File: Data Structures and Algorithms GFG/Chapter-6 Arrays/Topic-7 Remove Duplicates from a Sorted Array/Program_2.c
/*C Program to Remove Duplicates from a Fixed Sized
Array(Heap Allocation) Sorted in Non-Decreasing Order*/

#include<stdio.h>
#include<stdlib.h>
#include<limits.h>

/*Decalaring Global Variables*/
long long signed int capacity, size;
int *arr;

/*Function to Create a Fixed Sized
Array using Heap Allocation*/
short unsigned create_a_fixed_sized_array(void)
{
    /*User Input for capacity*/
    printf("Enter the Capacity of Array:");
    scanf("%lld", &capacity);
    /*User is advised to enter a Non-Negative Integer that is
    greater than or equal to 0 and less than (2 raised to power 63)
    as capacity otherwise the result will not be as predicted*/

    /*Checking for bounds for capacity*/
    if(capacity>LLONG_MAX || capacity<0)
    {
        printf("%lld is an Invalid Input for Capacity!\n", capacity);
        return 0;
    }

    /*User Input for size*/
    printf("Enter the Size of Array:");
    scanf("%lld", &size);
    /*User is advised to enter a Non-Negative Integer that is
    greater than or equal to 0 and less than (2 raised to power 63)
    as size otherwise the result will not be as predicted*/

    /*Checking for bounds for size*/
    if(size>LLONG_MAX || size<0)
    {
        printf("%lld is an Invalid Input for size!\n", size);
        return 0;
    }

    /*Comparison of size and capacity*/
    if(size>capacity)
    {
        printf("Size can't be greater than Capacity!\n");
        return 0;
    }

    /*Allocating Memory to Array*/
    arr=(int *)malloc(capacity*sizeof(int));

    /*Checking for Memory Allocation*/
    if(arr==NULL)
    {
        printf("Memory Allocation Failed!\n");
        return 0;
    }

    /*Taking the Elements of Array as User Input*/
    printf("Enter %lld Elements for Array:\n", size);
    for(long long unsigned int i=0; i<size; i++)
    {
        printf("Enter Element %lld of Array:", i+1);
        scanf("%d", &arr[i]);
    }
    return 1;
}

/*Function to Display Elements of Array: Display Function*/
void display(void)
{
    /*Displaying the Elements of Array: Traversing an Array*/
    printf("The Elements of Array are:\n");
    for(long long signed int i=0; i<size; i++)
    printf("Element %lld of Array is:%d\n", i+1, i[arr]);
}

/*Function to Check if an Array is Sorted 
in Non-Deceasing Order or not*/
short unsigned int check_for_sorted()
{
    for(long long signed int i=1; i<size; i++)
    {
        if(arr[i]<arr[i-1])
        return 0;
    }
    return 1; 
}

/*Function to Remove Duplicates from an 
Array Sorted in Non-Decreasing Order*/
void remove_duplicates()
{
    int *temp=(int *)malloc(size*(sizeof(int)));
    long long signed int result=1;
    temp[0]=arr[0];
    for(long long signed int i=1; i<size; i++)
    {
        if(temp[result-1]!=arr[i])
        {
            temp[result]=arr[i];
            result++;
        }
    }
    for(long long signed int j=1; j<result; j++)
    arr[j]=temp[j];
    size=result;
}

/*Driver Function(Code)*/
int main()
{
    if(!create_a_fixed_sized_array())
    printf("Array Creation Failed!");
    else
    {
        display();
        if(!check_for_sorted())
        printf("The Array is Not Sorted in Non-Decreasing Order!");
        else
        {
            printf("The Array is Sorted in Non-Decreasing Order!\n");
            remove_duplicates();
            printf("After Removing Duplicates, the Array becomes:\n");
            display();
        }
    }
    return 0;
}

File: Data Structures and Algorithms GFG/Chapter-6 Arrays/Topic-7 Remove Duplicates from a Sorted Array/Program_2.java
/*Java Program to Remove Duplicates from a Fixed Sized
Array(Heap Allocation) Sorted in Non-Decreasing Order*/

import java.util.*;

/*Creating an outer class*/
class Program_2{

    /*Creating an inner class*/
    class Implement_Fixed_Sized_Array{

        /*Data Members of class: int and int []*/
        int capacity, size, position;
        int []arr;
        int []temp;
        int element;

        /*Defining readInput Function*/
        void readInput(Scanner sc)
        {
            System.out.println("Enter the Capacity of Array:");
            capacity=sc.nextInt();
            System.out.println("Enter the Size of Array:");
            size=sc.nextInt();
        }

        /*Member Function of class:
        to Create a Fixed Sized
        Array using Heap Allocation*/
        boolean create_a_fixed_sized_array(Scanner sc)
        {
            readInput(sc);
            /*User is advised to enter a Non-Negative Integer that is
            greater than or equal to 0 and less than (2 raised to power 63)
            as capacity and size otherwise the result will not be as predicted*/

            /*Checking for bounds for capacity*/
            if(capacity>Integer.MAX_VALUE || capacity<0)
            {
                System.out.println(capacity+" is an Invalid Input for Capacity!");
                return false;
            }

            /*Checking for bounds for size*/
            if(size>Integer.MAX_VALUE || size<0)
            {
                System.out.println(size+" is an Invalid Input for Size!");
                return false;
            }

            /*Comparison of size and capacity*/
            if(size>capacity)
            {
                System.out.println("Size can't be greater than Capacity!");
                return false;
            }

            /*Allocating Memory to Array*/
            try{
                arr=new int[capacity];
            }
            catch(OutOfMemoryError e){
                System.out.println("Memory Allocation Failed!");
            }

            /*Taking the Elements of Array as User Input*/
            System.out.println("Enter "+size+" Elements for Array:");
            for(int i=0; i<size; i++)
            {
                System.out.println("Enter Element "+(i+1)+" of Array:");
                arr[i]=sc.nextInt();
            }
            return true;
        }

        /*Member Function of class:
        to Display Elements of Array:
        Display Function*/
        void display()
        {
            /*Displaying the Elements of Array: Traversing an Array*/
            System.out.println("The Elements of Array are:");
            for(int i=0; i<size; i++)
            System.out.println("Element "+(i+1)+" of Array is:"+arr[i]);
        }

        /*Member Function of class:
        to Check if an Array is Sorted 
        in Non-Deceasing Order or not*/
        boolean check_for_sorted()
        {
            for(int i=1; i<size; i++)
            {
                if(arr[i]<arr[i-1])
                return false;
            }
            return true;
        }

        /*Member Function of class:
        to Remove Duplicates from an Array
        Sorted in Non-Decreasing Order*/
        void remove_duplicates()
        {
            int result=1;
            for(int i=1; i<size; i++)
            {
                if(arr[result-1]!=arr[i])
                {
                    arr[result]=arr[i];
                    result++;
                }
            }
            size=result;
        }
    }

    /*Driver Function(Code)*/
    public static void main(String[] args)
    {
        Scanner sc=new Scanner(System.in);
        Program_2 program_2=new Program_2();
        Program_2.Implement_Fixed_Sized_Array IFSA=program_2.new Implement_Fixed_Sized_Array();
        if(!IFSA.create_a_fixed_sized_array(sc))
        System.out.println("Array Creation Failed!");
        else
        {
            IFSA.display();
            if(!IFSA.check_for_sorted())
            System.out.println("The Array is Not Sorted in Non-Decreasing Order!");
            else
            {
                System.out.println("The Array is Sorted in Non-Decreasing Order!");
                IFSA.remove_duplicates();
                System.out.println("After Removing Duplicates, the Array becomes:");
                IFSA.display();
            }
        }
        sc.close();
    }
}




